src/Substructure.jl

```diff
@@ -0,0 +1,313 @@
+    return (result, p1, p2)
+
+end
+
+"""
+    delta_R(jet1, jet2) -> Float64
+
+Function to calculate the distance in the y-ϕ plane between two jets `jet1` and `jet2`
+
+# Arguments
+- `jet1`: The first jet.
+- `jet2`: The second jet.
+
+# Returns
+- Float64: The Euclidean distance in the y-ϕ plane for the two jets.
+"""
+function delta_R(jet1::PseudoJet, jet2::PseudoJet)
+    
+    eta1, phi1 = rapidity(jet1), phi(jet1)
+    eta2, phi2 = rapidity(jet2), phi(jet2)
+
+    d_eta = eta1 - eta2
+    d_phi = phi1 - phi2
+    d_phi = abs(d_phi) > π ? 2π - abs(d_phi) : d_phi
+    
+    return sqrt(d_eta^2 + d_phi^2)
+
+end
+
+# Function to calculate kt distance between two pseudojets
+function kt_distance(jet1::PseudoJet, jet2::PseudoJet, R = 1)
+    p1 = pt2(jet1)
+    p2 = pt2(jet2)
+    
+    d_R = delta_R(jet1, jet2)
+    return min(p1, p2) * (d_R^2 / R^2)
+end;
+
+"""
+    recluster(jet, clusterseq, rad, mtd) -> ClusterSequence
+
+Reclusters the constituents of a given jet `jet` with a different clustering method `mtd` and different jet radius `rad`.
+
+# Arguments
+- `jet`: The jet whose constituents are to be reclustered.
+- `clusterseq`: The cluster sequence from which the original jet is obtained.
+- `rad`: The new jet radius.
+- `mtd`: The new clustering method.
+
+# Returns
+- ClusterSequence: The new cluster sequence.
+"""
+function recluster(jet::PseudoJet, clusterseq::ClusterSequence, rad = 1.0, mtd = 0)
+    cons = constituents(jet, clusterseq)
+    new_clusterseq = jet_reconstruct(cons; p = mtd, R = rad, strategy = RecoStrategy.N2Plain)
+   
+    return new_clusterseq
+end
+
+function sort_jets!(event_jet_array)
+    jet_pt(jet) = pt2(jet)
+    sort!(event_jet_array, by = jet_pt, rev = true)
+    # println(event_jet_array)
+end
+
+function join(jets::Vector{PseudoJet})
+    final = PseudoJet(0.0, 0.0, 0.0, 0.0)
+    
+    for jet in jets
+        final = final + jet
+    end
+
+    final
+end
+
+# Defining suitable structures to be used for jet grooming and tagging
+"""
+    struct MassDropTagger
+
+Used for tagging jets that undergo mass drop, a common technique in jet substructure.
+
+Fields:
+- mu: Maximum allowed mass ratio for a jet to pass tagging
+- y: Minimum kT distance threshold for parent separation
+"""
+
+mutable struct MassDropTagger
+    mu::Float64
+    y::Float64
+end
+
+"""
+    struct SoftDropTagger
+
+Applies a soft-drop condition on jets, trimming away soft, wide-angle radiation.
+
+Fields:
+- zcut: Minimum allowed energy fraction for subjets
+- b: Angular exponent controlling soft radiation suppression
+"""
+
+mutable struct SoftDropTagger
+    zcut::Float64
+    b::Float64
+end
+
+"""
+    struct Filter
+
+Filters jets based on radius and number of hardest subjets, reducing contamination.
+
+Fields:
+- filterRadius: Radius parameter to recluster subjets
+- numHardestJets: Number of hardest jets to retain in the filtered result
+"""
+
+mutable struct Filter
+    filterRadius::Float64 
```

Comment:
Avoid camel case for variable names in this module - I think `R` is the only exception (mea culpa). So `filter_radius` and `num_hardest_jets`.